numpy/core/function_base.py

```diff
@@ -123,25 +123,47 @@ def linspace(start, stop, num=50, endpoint=True, retstep=False, dtype=None,
     if num < 0:
         raise ValueError("Number of samples, %s, must be non-negative." % num)
     div = (num - 1) if endpoint else num
+    # Make sure one can use variables that have an __array_interface__, gh-6634
+    start = asanyarray(start)
+    stop  = asanyarray(stop)
+
+    # check if dealing with datetime64 or timedelta64
+    is_datelike = start.dtype.kind in 'Mm'
+    if is_datelike:
+        # For datetime/timedelta, dtype must be specified
+        # start and stop will first be cast to this dtype.
+        # Warning: dtype must have high enough precision so
+        # that it is at all possible to draw num equispaced
+        # samples.
+        assert dtype is not None, "linspace with {} objects cannot be called "\
+            "with dtype=None".format(start.dtype)
+        start = start.astype(dtype)
+        stop = stop.astype(dtype)
+        # y cannot be cast to datetime64/timedelta64, but left
+        # as int64, since later on we will do y * delta, and
+        # <m8[s] * <m8[s] is not allowed.
+        delta = stop - start
+        y = _nx.arange(0, num).reshape((-1,) + (1,) * ndim(delta))
 
-    # Convert float/complex array scalars to float, gh-3504
-    # and make sure one can use variables that have an __array_interface__, gh-6634
-    start = asanyarray(start) * 1.0
-    stop  = asanyarray(stop)  * 1.0
+    else:
+        # Convert float/complex array scalars to float, gh-3504
+        start = start * 1.0
+        stop  = stop  * 1.0
+        dt = result_type(start, stop, float(num))
+        delta = stop - start
+        y = _nx.arange(0, num, dtype=dt).reshape((-1,) + (1,) * ndim(delta))
 
-    dt = result_type(start, stop, float(num))
-    if dtype is None:
-        dtype = dt
+        if dtype is None:
+            dtype = dt
 
-    delta = stop - start
-    y = _nx.arange(0, num, dtype=dt).reshape((-1,) + (1,) * ndim(delta))
     # In-place multiplication y *= delta/div is faster, but prevents the multiplicant
     # from overriding what class is produced, and thus prevents, e.g. use of Quantities,
-    # see gh-7142. Hence, we multiply in place only for standard scalar types.
-    _mult_inplace = _nx.isscalar(delta)
+    # see gh-7142. Hence, we multiply in place only for standard scalar types,
+    # excepting datetime64 and timedelta64
+    _mult_inplace = _nx.isscalar(delta) and (not is_datelike)
```

Comment:
Why is this necessary?

Reply:
Because we cannot do `y *= delta` when `y` is `int` and `delta` is `timedelta64`.